core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+  public void testCommitConflict() {
+    Transaction txn = table.newTransaction();
+
+    txn.newFastAppend().appendFile(FILE_A).apply();
+    table.newFastAppend().appendFile(FILE_B).commit();
+
+    AssertHelpers.assertThrows("Should failed due to conflict",
+        IllegalStateException.class, "last operation has not committed", txn::commitTransaction);
+
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    AppendFiles appendFiles = table.newFastAppend().appendFile(FILE_C);
+    appendFiles.apply();
+    table.newFastAppend().appendFile(FILE_D).commit();
+    appendFiles.commit();
+
+    ManifestFile manifestFile = table.currentSnapshot().manifests().stream()
+        .filter(manifest -> manifest.snapshotId() == table.currentSnapshot().snapshotId())
+        .collect(Collectors.toList()).get(0);
+
+    for (ManifestEntry entry : ManifestFiles.read(manifestFile, table.io(),
+        table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_C.path())) {
+        Assert.assertEquals(table.currentSnapshot().sequenceNumber(), entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testConcurrentCommit() throws InterruptedException {
+    ExecutorService threadPool = Executors.newFixedThreadPool(4);
+    List<Callable<Void>> tasks = new ArrayList<>();
+
+    Callable<Void> write1 = () -> {
+      Transaction txn = table.newTransaction();
+      txn.newFastAppend().appendFile(FILE_A).commit();
+      txn.commitTransaction();
+      return null;
+    };
+
+    Callable<Void> write2 = () -> {
+      Transaction txn = table.newTransaction();
+      txn.newAppend().appendFile(FILE_B).commit();
+      txn.commitTransaction();
+      return null;
+    };
+
+    Callable<Void> write3 = () -> {
+      Transaction txn = table.newTransaction();
+      txn.newDelete().deleteFile(FILE_A).commit();
+      txn.commitTransaction();
+      return null;
+    };
+
+    Callable<Void> write4 = () -> {
+      Transaction txn = table.newTransaction();
+      txn.newOverwrite().addFile(FILE_D).commit();
+      txn.commitTransaction();
+      return null;
+    };
+
+    tasks.add(write1);
+    tasks.add(write2);
+    tasks.add(write3);
+    tasks.add(write4);
+    threadPool.invokeAll(tasks);
+    threadPool.shutdown();
+
+    Assert.assertEquals(4, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testRollBack() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_B).commit();
+
+    Assert.assertEquals(2, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    table.manageSnapshots().rollbackTo(snapshotId).commit();
+
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testMultipleTxnOperations() {
```

Comment:
This doesn't really test much because all of the transactions do not conflict and only contain one operation. I think there should be a few test cases (separate methods):

1. A transaction with 1 operation
2. A transaction with 2 operations
3. A transaction with an operation that doesn't change the sequence number (like expiration)
4. A transaction with 1 operation that must retry due to conflict

Reply:
Done.